**IMPORT DEPENDENCIES**

In [ ]:
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

**LAODING DATA**

In [ ]:
import pandas as pd
df = pd.read_csv('/content/car data.csv')

**DATA PREPROCESSING**

In [ ]:
df

Car_Name  Year  Selling_Price  Present_Price  Driven_kms Fuel_Type  \
0       ritz  2014           3.35           5.59       27000    Petrol   
1        sx4  2013           4.75           9.54       43000    Diesel   
2       ciaz  2017           7.25           9.85        6900    Petrol   
3    wagon r  2011           2.85           4.15        5200    Petrol   
4      swift  2014           4.60           6.87       42450    Diesel   
..       ...   ...            ...            ...         ...       ...   
296     city  2016           9.50          11.60       33988    Diesel   
297     brio  2015           4.00           5.90       60000    Petrol   
298     city  2009           3.35          11.00       87934    Petrol   
299     city  2017          11.50          12.50        9000    Diesel   
300     brio  2016           5.30           5.90        5464    Petrol   

    Selling_type Transmission  Owner  
0         Dealer       Manual      0  
1         Dealer       Manual      0  
2         Dealer       Manual      0  
3         Dealer       Manual      0  
4         Dealer       Manual      0  
..           ...          ...    ...  
296       Dealer       Manual      0  
297       Dealer       Manual      0  
298       Dealer       Manual      0  
299       Dealer       Manual      0  
300       Dealer       Manual      0  

[301 rows x 9 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Car_Name       301 non-null    object 
 1   Year           301 non-null    int64  
 2   Selling_Price  301 non-null    float64
 3   Present_Price  301 non-null    float64
 4   Driven_kms     301 non-null    int64  
 5   Fuel_Type      301 non-null    object 
 6   Selling_type   301 non-null    object 
 7   Transmission   301 non-null    object 
 8   Owner          301 non-null    int64  
dtypes: float64(2), int64(3), object(4)
memory usage: 21.3+ KB


In [ ]:
df.describe()

Year  Selling_Price  Present_Price     Driven_kms       Owner
count   301.000000     301.000000     301.000000     301.000000  301.000000
mean   2013.627907       4.661296       7.628472   36947.205980    0.043189
std       2.891554       5.082812       8.642584   38886.883882    0.247915
min    2003.000000       0.100000       0.320000     500.000000    0.000000
25%    2012.000000       0.900000       1.200000   15000.000000    0.000000
50%    2014.000000       3.600000       6.400000   32000.000000    0.000000
75%    2016.000000       6.000000       9.900000   48767.000000    0.000000
max    2018.000000      35.000000      92.600000  500000.000000    3.000000

In [ ]:
df.isnull().sum()# checking null values

Car_Name         0
Year             0
Selling_Price    0
Present_Price    0
Driven_kms       0
Fuel_Type        0
Selling_type     0
Transmission     0
Owner            0
dtype: int64

In [ ]:
print(df.columns)

Index(['Car_Name', 'Year', 'Selling_Price', 'Present_Price', 'Driven_kms',
       'Fuel_Type', 'Selling_type', 'Transmission', 'Owner'],
      dtype='object')


In [ ]:
df.	Fuel_Type[df.	Fuel_Type == 'Diesel'] = 0
df.	Fuel_Type[df.	Fuel_Type == 'Petrol'] = 1

In [ ]:
df

Car_Name  Year  Selling_Price  Present_Price  Driven_kms Fuel_Type  \
0       ritz  2014           3.35           5.59       27000         1   
1        sx4  2013           4.75           9.54       43000         0   
2       ciaz  2017           7.25           9.85        6900         1   
3    wagon r  2011           2.85           4.15        5200         1   
4      swift  2014           4.60           6.87       42450         0   
..       ...   ...            ...            ...         ...       ...   
296     city  2016           9.50          11.60       33988         0   
297     brio  2015           4.00           5.90       60000         1   
298     city  2009           3.35          11.00       87934         1   
299     city  2017          11.50          12.50        9000         0   
300     brio  2016           5.30           5.90        5464         1   

    Selling_type Transmission  Owner  
0         Dealer       Manual      0  
1         Dealer       Manual      0  
2         Dealer       Manual      0  
3         Dealer       Manual      0  
4         Dealer       Manual      0  
..           ...          ...    ...  
296       Dealer       Manual      0  
297       Dealer       Manual      0  
298       Dealer       Manual      0  
299       Dealer       Manual      0  
300       Dealer       Manual      0  

[301 rows x 9 columns]

**MACHINE LEARNIG**

In [ ]:
X = df[['Year', 'Present_Price', 'Driven_kms', 'Fuel_Type', 'Selling_type', 'Transmission', 'Owner']]
y = df['Selling_Price']


In [ ]:
X = pd.get_dummies(X, columns=['Fuel_Type', 'Selling_type', 'Transmission'], drop_first=True)

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


**Linear Regression is a regression algorithm that models the relationship between the input features and the target variable as a linear equation.
The goal is to predict a continuous numerical output in this case it is selling Prize.**

In [ ]:
model = LinearRegression()

In [ ]:
model.fit(X_train, y_train)

LinearRegression()

In [ ]:
y_pred = model.predict(X_test)

In [ ]:
mae = mean_absolute_error(y_test, y_pred) #This metric represents the average absolute difference between the actual and predicted values.
print(f'Mean Absolute Error: {mae}')

Mean Absolute Error: 1.216374019333193


In [ ]:
mse = mean_squared_error(y_test, y_pred) #This metric represents the average squared difference between the actual and predicted values
print(f'Mean Squared Error: {mse:.2f}')

Mean Squared Error: 3.48


In [ ]:
r2 = r2_score(y_test, y_pred) #This metric indicates the proportion of variance in the target variable that is explained by the model.
print(f'R-squared: {r2:.2f}')


R-squared: 0.85


**A lower MAE indicates better model performance.**

**R² values range from 0 to 1, with higher values indicating better fit.**

In [ ]:
coefficients = pd.DataFrame({'Feature': X_train.columns, 'Coefficient': model.coef_})
print("Model Coefficients:")
print(coefficients)

Model Coefficients:
                   Feature  Coefficient
0                     Year     0.352925
1            Present_Price     0.429070
2               Driven_kms    -0.000006
3                    Owner    -0.905018
4              Fuel_Type_1    -1.795750
5            Fuel_Type_CNG    -2.534904
6  Selling_type_Individual    -1.191642
7      Transmission_Manual    -1.640074


In [ ]:
train_predictions = model.predict(X_train)

In [ ]:
mae_train = mean_absolute_error(y_train, train_predictions)
mse_train = mean_squared_error(y_train, train_predictions)
r2_train = r2_score(y_train, train_predictions)
print(f'MAE on Training Set: {mae_train:.2f}')
print(f'MSE on Training Set: {mse_train:.2f}')
print(f'R-squared on Training Set: {r2_train:.2f}')

MAE on Training Set: 1.17
MSE on Training Set: 2.95
R-squared on Training Set: 0.89


**On average, the model's predictions deviate by 1.17 units from the actual selling prices. The R-squared value of 0.89 indicates that the model explains approximately 89% of the variance in the selling prices**

In [ ]:
new_data = pd.DataFrame({
    'Car_Name': ['ciaz'],  # provide input for the car whose selling prize is to be predicted
    'Year': [2017],
    'Present_Price': [9.85],
    'Driven_kms': [6900],
    'Fuel_Type': ['Petrol'],
    'Selling_type': ['Dealer'],
    'Transmission': ['Manual'],
    'Owner': [0]
})

new_data_processed = pd.get_dummies(new_data, columns=['Fuel_Type', 'Selling_type', 'Transmission'], drop_first=True)

missing_cols = set(X_train.columns) - set(new_data_processed.columns)
for col in missing_cols:
    new_data_processed[col] = 0

new_data_processed = new_data_processed[X_train.columns]

new_predictions = model.predict(new_data_processed)

predicted_data = pd.DataFrame({  # this will print the car name with its predicted selling prize
    'Car_Name': new_data['Car_Name'],
    'Predicted_Price': new_predictions
})

# Display the predictions
print("Predictions for new data:")
print(predicted_data)

Predictions for new data:
  Car_Name  Predicted_Price
0     ciaz        10.382119
